File: Stack/C++/StackList.cpp
#include <iostream>
#include "StackList.h"

using namespace std;

template <class T>
Node<T>::Node(){

}

template <class T>
Node<T>::Node(T value){
    this->data = value;
    this->next = NULL;
}

template <class T>
StackList<T>::StackList()
{
    this->head = NULL;
    this->top = NULL;
    this->length = 0;
}

template <class T>
void StackList<T>::push(T value){
    Node<T> *newNode = new Node<T>(value);
    if(head == NULL){
        head = newNode;
        top = newNode;
    }
    else{
        top->next = newNode;
        top = newNode;
    }
    length++;
}

template <class T>
void StackList<T>::pop(){
    Node<T> *current = head;
    while(current->next != NULL){
        current = current->next;
    }
    top = NULL;
    current->next = NULL;
    top = current;
    length--;
}

template <class T>
void StackList<T>::print(){
    Node<T> *current = head;
    while(current != NULL){
        cout << current->data << " ";
        current = current->next;
    }
    cout << endl;
}

template <class T>
T StackList<T>::peek(){
    return top->data;
}

template <class T>
int StackList<T>::getLength(){
    return length;
}

File: Stack/Java/StackList.java
public class StackList<T>{
    class Node<T>{
        T data;
        Node<T> next;
        Node(){}
        Node(T value){
            this.data = value;
            this.next = null;
        }
    }
    private Node<T> head,top;
    private int length;
    StackList(){
        this.head = null;
        this.top = null;
        this.length = 0;
    }
    void push(T value){
        Node<T> newNode = new Node<>(value);
        if(head == null){
            head = newNode;
            top = newNode;
        }
        else {
            top.next = newNode;
            top = newNode;
        }
        length++;
    }
    void pop(){
        if(head == null){
            System.out.println("Stack Underflow");
            return;
        }
        Node<T> current = head;
        while (current.next != null)
            current = current.next;
        top = null;
        current.next = null;
        top = current;
        length--;
    }
    void print(){
        Node<T> current = head;
        while (current != null){
            System.out.print(current.data + " ");
            current = current.next;
        }
        System.out.println();
    }
    T peek(){
        return top.data;
    }
    int getLength(){
        return length;
    }
}
